### DISCUSSION TITLE
RGB LEDs vs. having 10+ monochromatic light sources

### DISCUSSION THREAD

**Discussion by sgbaird** [sparks-baird/self-driving-lab-demo#52] [Demo extensions and design]
An array of RGB LEDs + a spectrometer might not be the best match-up. RGB LEDs + a sensor designed for a single color value (e.g. [RGB Color Sensor with IR filter and White LED - TCS34725](https://www.adafruit.com/product/1334)) or monochromatic lasers (e.g. [Laser Diode - 5mW 650nm Red](https://www.adafruit.com/product/1054)) + the 10-channel spectrometer might be a better match-up. Tunable laser(s) probably too expensive.

See also https://www.adafruit.com/product/3595

Will see once the hardware comes in and I start testing.

**Comment by sgbaird** [comment]
https://www.amazon.com/Chanzon-100pcs-colors-Emitting-Assorted/dp/B01AUI4W5U?th=1#

**Comment by sgbaird** [comment]
Submitted a CrowdSupply application and contacted Pimoroni.

Message to Pimoroni:

> I'm looking to make a low-cost, open-source self-driving (autonomous) lab demo while retaining the main principles of closed-loop feedback laboratory automation. https://github.com/sparks-baird/self-driving-lab-demo
> 
> Blinkt! is amazing, but since each node is a set of RGB LEDs, in the end I really only get 3 unique wavelength types (maybe with some manufacturer variability), which is not quite as high-dimensional or interesting as I was hoping.
> 
> So, I'm looking all over for something that is an addressable array of LEDs of different wavelengths (~10 preferable) around the visible spectrum. https://www.arborsci.com/products/led-array is along the right lines, but too big, too expensive, and not addressable (flip switches by hand). https://www.amazon.com/Chanzon-100pcs-colors-Emitting-Assorted/dp/B01AUI4W5U?th=1 is great because it has LEDs spanning a range of wavelengths, but these are designed for a breadboard. DigiKey has a wide selection as well, but the LEDs would need to be sourced individually, probably from different suppliers. Even if I identified a cohesive set of LEDs, this requires additional user expertise (soldering), and all options mentioned so far would also require a custom Python program (not as big of a deal for me).
> 
> None of this would really be a deal-breaker, except that I'm trying to make it so the self-driving lab demo can be purchased in one or two online shopping carts for less than $100, takes up less than a square foot of desk space, and can be set up in less than an hour with no other tools than what gets shipped (and preferably no soldering), in a scalable way so that if 100's of people with virtually no electronics experience and basic coding could get it up and running within those requirements.
> 
> Unless you're aware of something that meets these specifications, I'd love to get this going via e.g. CrowdSupply or hear any suggestions for how to make something like this available to myself and many others.

**Comment by sgbaird** [comment]
Some relevant questions and resources to "how do I get a tunable wavelength light source" or similar:
- https://physics.stackexchange.com/questions/375649/what-are-the-other-ways-to-pass-particular-wavelength-of-light-without-using-opt?rq=1
- https://physics.stackexchange.com/questions/100682/what-is-the-best-way-to-collimate-light-emitted-by-a-led?rq=1
- https://physics.stackexchange.com/questions/197502/variable-wavelength-led
- https://en.wikipedia.org/wiki/Monochromator
- https://www.quora.com/How-do-I-change-the-wavelength-of-a-LED-bulb-Can-it-be-changed-by-changing-the-voltage
- https://electronics.stackexchange.com/questions/223723/led-with-variable-wavelength

Instructables or similar:
- https://www.instructables.com/Multiwavelength-LED-Light-Source/
- https://www.thepulsar.be/article/building-a-prism-monochromator/
- https://www.youtube.com/watch?v=ypFxMRxMtmQ
- https://hackaday.com/2014/11/22/creating-a-scanning-monochromator/
- https://www.instructables.com/A-simple-DIY-spectrophotometer/
- https://www.instructables.com/DIY-Low-Cost-Spectrometer/
- https://opg.optica.org/ao/fulltext.cfm?uri=ao-11-2-469&id=163839

Other products:
- https://www.superbrightleds.com/moreinfo/led-strips-and-bars/5m-single-color-led-strip-light-eco-series-tape-light-12v24v-ip20/5141/
- https://www.amazon.com/Correction-Colored-Overlays-Transparency-Plastic/dp/B07WZM6TDT/ref=asc_df_B07WZM6TDT/

Expensive products:
- https://www.prizmatix.com/FC/Multi-Wavelength-Fiber-Coupled-LEDs.aspx
- https://tech-led.com/en/marubeni-opto/multi-wavelength-leds/
- https://gamma-sci.com/products/spectralled-tunable-light-sources/
- https://www.sciencetech-inc.com/shop/category/modular-spectrometers-monochromators-and-spectrographs-small-series-97
- https://www.photonics.com/Products/Digital_Mini-Chrom_Monochromators/pr63219
- https://www.spectralproducts.com/Monochromators/DK480
- https://www.edmundoptics.com/f/manual-mini-chrom-monochromators/11508/
- https://www.edmundoptics.com/p/linear-variable-vis-bandpass-filter/30623

Theory:
- https://zeiss-campus.magnet.fsu.edu/print/lightsources/leds-print.html
- https://www.quora.com/How-does-the-wavelength-of-light-change-when-passed-through-a-coloured-transparent-material (while applicable, a good bandpass filter is probably expensive) 
- https://physics.stackexchange.com/questions/566687/why-the-color-of-light-observed-dark-almost-no-light-from-section-of-overlappi

**Comment by sgbaird** [comment]
xref-ing the 10-channel wavelengths with [color descriptors](https://www.everredtronics.com/Led.Color.Visible.color.html):
```
415 nm – Violet
445 nm – Indigo or Blue
480 nm Blue or Blue-Green
515 nm Blue-Green or Green
555 nm Green or Yellow-Green 
590 nm – Yellow-Green or Yellow
630 nm Orange or Orange-Red or Red
680 nm - Red
Clear ?
Near IR – Shortwave NIR or Longwave NIR
```

**Comment by sgbaird** [comment]
Dotstar wavelengths are certainly narrow:
![image](https://user-images.githubusercontent.com/45469701/177462205-4cd4a27a-5063-4555-97c6-a98b8cb428f2.png)
([source](https://cdn-shop.adafruit.com/product-files/2343/SK9822_SHIJI.pdf), mentioned by [adafruit_support_bill](https://forums.adafruit.com/memberlist.php?mode=viewprofile&u=10091) in https://forums.adafruit.com/viewtopic.php?f=8&t=192420)

**Comment by sgbaird** [comment]
Here's an example of non-linear, non-monotonic behavior:
<img src=https://user-images.githubusercontent.com/45469701/177462657-34067c4e-fd2e-4cf8-8755-595cf72c213f.png width=300>
(same [source](https://cdn-shop.adafruit.com/product-files/2343/SK9822_SHIJI.pdf))

based on:
<img src=https://user-images.githubusercontent.com/45469701/177462779-0fc6cd81-1920-4164-b789-c8fd8ff05801.png width=300>

using forward voltage as a parameter could introduce a non-linearity into the task. For example, multi-objective optimization of maximizing luminescence, color matching, and minimizing total power (granted forward voltage is likely not controllable with Blinkt!, at which point it might be easier to use the separate wavelengths instead). This is of course assuming that the inflection point isn't going to lead to complete device failure with normal, repeated use.

If I'm trying to maximize luminescence, then I might need to monitor the temperature of the board. However, it's also probable that the influence of temperature on wavelength will be low relative to the spectrometer resolution; meanwhile, the influence of temperature on flux/efficiency will still probably be there.

This doesn't necessarily imply multiple local optima, however.

**Comment by sgbaird** [comment]
Considering [PCBWay](https://www.pcbway.com/) for assembled PCB service. Probably should make a breadboard prototype with LEDs from Amazon first for verification.

**Comment by sgbaird** [comment]
Liam from PCBWay generously offered free prototyping for this project.

**Comment by sgbaird** [comment]
For NeoPixel, it's different data https://forums.adafruit.com/viewtopic.php?f=47&t=66146.

- https://github.com/sparks-baird/self-driving-lab-demo/tree/main/src/self_driving_lab_demo/data
- Used the [Avantes Software](https://www.avantes.com/support/theoretical-background/introduction-to-spectroscopy-software/) (Demo mode) to retrieve the CSVs
- haven't dealt explicitly with FHWM's in simulation for the sensor readings (https://brainder.org/2011/08/20/gaussian-kernels-convert-fwhm-to-sigma/)

**Comment by jakobzeitler** [comment]
Hugely useful resources, thanks for sharing!

**Reply by sgbaird** [reply]
Glad it's helpful!
